Java Streaming

josephlewis42 opened the pull request:
Use the BigQuery Java streaming client rather than batching files locally. Bumps version to 4.0.0

The streaming client allows the plugin to be ephemeral, removes the complexity around rotating files, checking for jobs, etc. and adds the ability to do instant queries/alerting on the logs. Batches can also be done by time period, total pending size of records or total pending record count.

The downside is a cost of $0.05 per GB of successfully inserted rows. Rows under 1KB are rounded to 1KB. Uploads as large batch files are free.

If this is merged it should close the following issues: 

 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/36
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/33
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/20
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/19
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/10
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/9
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/4

And the following PRs:

 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/32
 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/25

And possibly the issue: 

 * https://github.com/logstash-plugins/logstash-output-google_bigquery/issues/34

rosbo commented:
Are you looking at why the CI build is failing?

rosbo commented on README.md:
````diff
@@ -29,6 +29,7 @@ Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/log
 - Install dependencies
 ```sh
 bundle install
+rake install_jars
````
You should run `bundle exec rake install_jars` if you want the rake command to run in the context of the current bundle.

josephlewis42 replied:
Perfect! that's probably why my dependency conflicts weren't caught before Travis tried running

rosbo commented on lib/logstash/outputs/bigquery/streamclient.rb:
```diff
@@ -0,0 +1,120 @@
+            return com.google.cloud.bigquery.BigQueryOptions.getDefaultInstance().getService()
+          end
+
+          # TODO: set User-Agent
```
Consider creating an issue in GH and linking to this todo

josephlewis42 replied:
Will do, I'll add an action item to add it once this branch gets merged. I think the existing solution derives a UA string from the hard-coded application name and version. But it doesn't really fit with the ones we gave the beats.

rosbo commented on lib/logstash/outputs/bigquery/schema.rb:
```diff
@@ -0,0 +1,86 @@
+        include_package 'com.google.cloud.bigquery'
+
+        # Converts a CSV schema or JSON schema into a BigQuery Java Schema.
+        def self.parse_csv_or_json(csv_schema, json_schema)
```
Where is this code used?

josephlewis42 replied:
This is used in google_bigquery.rb:177, I'm calling it at register so we fail early if the user inputs an invalid schema.

rosbo replied:
Thanks. The diff was collapsed so Ctrl + F failed me :)

josephlewis42 commented:
@rosbo I got the builds working except for the 5.6 branch which is a known compatibility issue between Travis and JAR vendoring: https://github.com/logstash-plugins/logstash-input-kafka/issues/241

josephlewis42 commented:
@robbavey thanks for your time! I'll fix these up tomorrow. Jake Landis just pointed me at the directions to "gradelize" in another PR so I'll get that done in here too.

Would you mind telling me the correct way to re-generate the `index.asciidoc`? Or do you hand-update after the initial generation? I saw an old forum post from ~2015 for a sub-command off the main Logstash executable but it doesn't seem to exist anymore.

robbavey commented:
@josephlewis42 Typically, we will make the changes directly in the .asciidoc file. That could be done in a subsequent PR, but it would be good to have those changes in before the gem is published.

josephlewis42 commented:
@robbavey I think I've fix it all up now. I've manually merged in the new docs, updated the CHANGELOG, moved to gradle, fixed the couple of code suggestions you had and added another set of unit-tests around schema parsing.

Travis has been hanging on me but I did trigger it through my repo here in case it still hasn't kicked off by the time you review this: https://travis-ci.org/josephlewis42/logstash-output-google_bigquery/builds/372266110

Thanks for the help and let me know if there's anything else you'd like to see.

robbavey commented on lib/logstash/outputs/google_bigquery.rb:
```diff
@@ -141,539 +130,151 @@ class LogStash::Outputs::GoogleBigQuery < LogStash::Outputs::Base
+  config :service_account, validate: :string, obsolete: 'Use json_key_file or ADC instead.'
+  config :temp_file_prefix, validate: :string, obsolete: 'No longer used.'
+  config :temp_directory, validate: :string, obsolete: 'No longer used.'
 
```
Would you remind deprecating rather than obsoleting the 'no longer used' settings - I assume from your comments in the changelog that an existing pipeline using the `ADC` will continue to work with this implementation? Deprecating the settings enables the pipeline to continue working with warnings, until the config setting is marked `obsolete` at a later date, when using that setting will fail the pipeline.

josephlewis42 replied:
Sure thing! I'll mark all  of them except the `key_path` and `key_password`. If someone was using ADC those shouldn't be defined at all.

If they were using the old key style, I'd rather that blow up on them than silently falling back to credentials that might be sitting in their environment variables that they explicitly overrode with the P12 style key file. How does that sound?

robbavey replied:
That makes sense - thanks!

robbavey commented:
@josephlewis42 One more thing - can you update the `changelog.md` to reflect `obsolete` vs `deprecated` fields, and update the docs to keep those fields in, but mark them as such - mark deprecated fields with something like `This option is deprecated, and is no longer used`, and for the obsolete fields, an explanation that the field is obsolete, and has been replaced with the `json_key_file` /`Application Default Credentials`.

Thanks!

josephlewis42 commented:
@robbavey sure thing! I've added it in to the `changelog.md` and the `asciidoc` along with pointers about how to upgrade and behavioral differences due to the changed fields.

robbavey commented:
@josephlewis42 I'm not sure if you have merge/publish privileges - if not, let me know and I'll push the necessary buttons

robbavey commented:
@josephlewis42 Go ahead and squash and merge at your leisure - I can kick off the publish bot once the master branch has built correctly.

josephlewis42 commented:
Awesome, all done and things are passing. @robbavey thanks for all the help! I'll close out the tickets with instructions about how to upgrade once I see the bot.

robbavey commented:
@josephlewis42 The gem is now published - I had to do a quick PR to resolve a missing dependency to enable publishing the gem, but that is done now and the newly released gem is up on rubygems.org.

Thank you for the contribution!
